src/Microsoft.AspNetCore.Sockets.Common/Internal/Formatters/ServerSentEventsMessageParser.cs

```diff
@@ -0,0 +1,208 @@
+            var end = examined;
+
+            while (!reader.End)
+            {
+                if (ReadCursorOperations.Seek(start, end, out var lineEnd, ByteLF) == -1)
+                {
+                    if(end != buffer.End)
+                    {
+                        return ParseResult.Incomplete;
+                    }
+                    throw new FormatException("Expected a '\r\n' line ending");
+                }
+
+                lineEnd = buffer.Move(lineEnd, 1);
+                var line = ConvertBufferToSpan(buffer.Slice(start, lineEnd));
+                reader.Skip(line.Length);
+
+                if(!(line.Length > 1))
+                {
+                    throw new FormatException("Error parsing data from event stream");
+                }
+
+                //Strip the \r\n from the span
+                line = line.Slice(0, line.Length - 2);
+
+                switch (_internalParserState)
+                {
+                    case InternalParseState.ReadMessageType:
+                        messageType = GetMessageType(line);
+                        start = lineEnd;
+                        _internalParserState = InternalParseState.ReadMessagePayload;
+                        consumed = lineEnd;
+
+                        //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                        if (reader.Peek() == ByteCR)
+                        {
+                            _internalParserState = InternalParseState.ReadEndOfMessage;
+                        }
+                        break;
+                    case InternalParseState.ReadMessagePayload:
+                        if (!StartsWithDataPrefix(line))
+                        {
+                            throw new FormatException("Expected the message prefix 'data: '");
+                        }
+
+                        //Slice away the 'data: '
+                        var newData = line.Slice(line.IndexOf(ByteSpace) + 1).ToArray();
+                        start = lineEnd;
+                        _data.Add(newData);
+
+                        //Peek into next byte. If it is a carriage return byte, then advance to the next state
+                        if (reader.Peek() == ByteCR)
+                        {
+                            _internalParserState = InternalParseState.ReadEndOfMessage;
+                        }
+
+                        consumed = lineEnd;
+                        break;
+                    case InternalParseState.ReadEndOfMessage:
+                        if (ReadCursorOperations.Seek(start, end, out lineEnd, ByteLF) == -1)
+                        {
+                            // The message has ended with \r\n\r
+                            return ParseResult.Incomplete;
+                        }
+
+                        //To check for the invalid case of there being more data after the frame end \r\n\r\n
+                        if (!reader.End)
```

Comment:
There could be another frame in the buffer already, so we can't assume this. We shouldn't need this check at all anyway, all we need to do is set `consumed` to the end of the current frame and let the caller call us again.